Add sources and documentation

rshym opened the pull request:
The script was developed on the private repo.
This PR adds all sources and readme for review.

rshym committed:
Introduce xt_imager.py

Based on xen-troops/rcar_flash.

Signed-off-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Ihor Usyk <ihor_usyk@epam.com>

rshym committed:
Add support for the internal TFTP server

This commit makes it possible to use `-t` without a path
to start the internal TFTP server to feed the image to the u-boot.

The root rights are required to open port 69 on the local host.
So you need to install `pyserial` and `tftpy` under sudo too.

There are two options available:
```
./xt_imager.py -t <path_to_external_tftp_server> <path_to_image_file>
```
or
```
sudo ./xt_imager.py -t -- <path_to_image_file>
```
Last option uses `--` to separate the image file from the `-t` option.

Signed-off-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Ihor Usyk <ihor_usyk@epam.com>

rshym committed:
Add the possibility to set IP in the u-boot

This commit allows us to handle situations when there is no `serverip`
or `ipaddr` in the u-boot environment or temporarily use other values.
So we add the possibility to specify `--serverip` for host and
`--ipaddr` for the board.
If not provided then u-boot will use settings already available in
the u-boot environment.

Signed-off-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Ihor Usyk <ihor_usyk@epam.com>

rshym committed:
Update README.md

Command line options are explained and examples are provided.

Signed-off-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Ihor Usyk <ihor_usyk@epam.com>

rshym committed:
Add support for .xz packed images.

Such archives reduce the filesize about 10x and can be
read incrementally, without unpacking of the full file.

There are two minor disadvantages:
- not significant increase of flashing time
  (for example, for one test case we see 5m7 vs 5m20);
- the file size of the full image size is not accessible
  before full unpack, so we can't print the percentage of
  the flashing progress.

Signed-off-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Ihor Usyk <ihor_usyk@epam.com>

iusyk commented on xt_imager.py:
```diff
@@ -0,0 +1,259 @@
+        for i in range(len(data)):
+            # check for zero only
+            if (data[i] != 0):
+                buffer_is_00_only = False
```
the 'break' looks reasonable here

rshym replied:
Thanks for pointing!
`break` is added.

iusyk commented on xt_imager.py:
```diff
@@ -0,0 +1,259 @@
+    chunk_filename = "chunk.bin"
+    chunk_size_in_bytes = 20*1024*1024
+
+    use_lzma = False
```
not required , the following code should work 
if str(args.image).endswith(".xz"):
        f = lzma.open(args.image)
        image_size = 0
    else:
        f = open(args.image, "rb")
all the rest is the same

rshym replied:
In the initial stages of development, it was just like you proposed.
But I decided to use a dedicated variable because I do not feel comfortable with string manipulations inside the loop, which (manipulation) is not required, and can be 'cached' inside bool.

iusyk replied:
then use just use_lxma

iusyk commented on xt_imager.py:
```diff
@@ -0,0 +1,259 @@
+    print("")
+
+    os.remove(out_fullname)
+    if use_lzma:
```
see my comment above regarding the usage of variable 'use_lzma', looks like it is not necessary

rshym replied:
Answered above.

iusyk commented on xt_imager.py:
```diff
@@ -0,0 +1,259 @@
+    baud = 115200
+
+    log.info(f"Using serial port {dev_name} with baudrate {baud}")
+    conn = serial.Serial(port=dev_name, baudrate=baud, timeout=20)
```
maybe I am wrong, but call of Serial(port=dev_name, baudrate=baud, timeout=20) opens the port, after that we close it and open again
i think, it is not necessary , it is enough to call Serial(port=dev_name, baudrate=baud, timeout=20) and that is it

rshym replied:
You are right! Thanks! I have cleaned this code.

iusyk reviewed (commented):
please, look at the comments, nothing critical but, i assume  - reasonable

rshym commented:
PR is reworked and re-tested.
The number of commits was reduced from 11 down to 6.
Comments are addressed.

lorc commented on xt_imager.py:
```diff
@@ -0,0 +1,166 @@
+    log.info(f"Using serial port {dev_name} with baudrate {baud}")
+    return serial.Serial(port=dev_name, baudrate=baud, timeout=20)
+
+def conn_wait_for(conn, expect: str):
```
Small nitpick, but:  you can just call `conn_wait_for_any(conn, [expect])` here and that's all. No need for duplicated code.

rshym replied:
Thanks, fixed.

lorc commented on xt_imager.py:
```diff
@@ -0,0 +1,166 @@
+    # send "newline char" to start further output on the new line
+    print("")
+
+    os.remove(out_fullname)
```
You need put this into `finally:` block, otherwise `chunk.bin` will be left in TFTP directory if there was an exception during flash process.

rshym replied:
Thanks, fixed.

lorc commented on xt_imager.py:
```diff
@@ -28,17 +30,53 @@ def main():
+
+class PYTFTPServer(object):
+    def __init__(self, folder):
+        self.tftp_server = tftpy.TftpServer(folder)
```
TftpServer provides this one cool feature:

>     dyn_file_func is a callable that takes a requested download
>     path that is not present on the file system and must return either a
>     file-like object to read from or None if the path should appear as not
>     found. This permits the serving of dynamic content.

This is why I suggested to use internal TFTP server in the first place.  I am not asking to use it right now, just FYI.

rshym replied:
I remember, but unfortunately, my attempt to implement this was unsuccessful, so I decided to use the current solution at least for now.

lorc commented on xt_imager.py:
```diff
@@ -135,21 +135,32 @@ def do_flash_image(args, tftp_root):
-        conn_wait_for(conn, f"Bytes transferred = {len(data)}")
-        conn_wait_for(conn, uboot_prompt)
+        buffer_is_00_only = True
+        for i in range(len(data)):
```
```
buffer_is_00_only = False
for x in data:
    if data != 0:
        break
else: # Pay attention - this "else" is for "for" loop, not for "if" expression. https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops
    buffer_is_00_only = True
```

rshym replied:
Thanks! I reworked this commit and got rid of the `buffer_is_00_only` variable.
`tftp` and `mw.b` are used inside this `for ... else ...` construction with corresponding comments.

lorc commented on xt_imager.py:
```diff
@@ -134,8 +143,14 @@ def do_flash_image(args, tftp_root):
-        data = f_img.read(chunk_size_in_bytes)
+    while True:
+        if use_lzma:
+            data = f_lzma.read(chunk_size_in_bytes)
```
So what is the difference? Python has a concept of "file-like object".  `lzma.open()` returns this type of object.

You can use `f_img` in both cases. Just initialize it with either `open` or `lzma.open`.

rshym replied:
Thanks, reworked.

lorc reviewed (commented):
There are few minor comments and one big question:

U-Boot supports LZMA decompression:

```
=> lzmadec
lzmadec - lzma uncompress a memory region

Usage:
lzmadec srcaddr dstaddr [dstsize]
```

Will it be more efficient to sent LZMA-compressed chunks and decompress them on U-Boot side?

rshym commented:
> Will it be more efficient to sent LZMA-compressed chunks and decompress them on U-Boot side?

I think that this is a VERY interesting idea! Thanks, I will implement it. Looks like we may get some time improvements with the slowest part of the script - TFTP transfer.

UPDATE. I implemented this feature, and it's a game-changer. Flashing time decreased almost 3x from 2m57 down to 1m14.
I have put PR into [WIP] state to make some additional tests, and will new commit to PR.

rshym commented:
Reworked. Successfully tested all scenarios (external/internal TFTP server, error hadnling, raw/compressed image).
Ready for re-review.

lorc reviewed (approved):
Thanks

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

rshym committed:
Compress data that is sent to u-boot

This commit compresses chunks before TFTP transfer and
decompresses them on the u-boot side.
This mechanism dramatically reduces the total time of
flashing, because the slowest part of the script is
TFTP transfer.
Compressing chunks we reduce the size of transferred
files from 20MB down to 4-6MB usually. And in the case
of empty blocks, filled with zeroes, we see compression
down to 3500 bytes.

Overall, we see a decrease in flashing time from 3
minutes down to almost 1m30 for a 1GB image.

Suggested-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Ihor Usyk <ihor_usyk@epam.com>

iusyk commented:
Reviewed-by: Ihor Usyk <ihor_usyk@epam.com>

rshym commented:
Reviewers are added.
Ready to be merged.
